Ironically, CI also fails with

```
C:/Users/runneradmin/AppData/Local/Temp/jl_Cwnmue/artifacts/b17bda08a19173572926f43a48aad5ef3d845e7c/extracted_files/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: error: export ordinal too large: 92817
collect2.exe: error: ld returned 1 exit status
```

--- a/src/PackageCompiler.jl
+++ b/src/PackageCompiler.jl
@@ -806,7 +806,7 @@ function get_extra_linker_flags(version, compat_level, soname)
     soname_arg = soname === nothing ? `` : `-Wl,-soname,$soname`
     rpath_args = rpath_sysimage()
 
-    extra = Sys.iswindows() ? `-Wl,--export-all-symbols` :
+    extra = Sys.iswindows() ? (VERSION >= v"1.11" ? `-Wl,--export-all-symbols` : ``) :
             Sys.isapple() ? `-fPIC $compat_ver_arg $current_ver_arg $rpath_args` :
             Sys.isunix() ? `-fPIC $soname_arg $rpath_args` :
                 error("unknown machine type, not windows, macOS not UNIX")

--- a/src/julia_init.c
+++ b/src/julia_init.c
@@ -8,17 +8,24 @@ JL_DLLEXPORT char *dirname(char *);
 #include <libgen.h>
 #endif
 
+#ifdef _OS_WINDOWS_
+#define DLLEXPORT __declspec(dllexport)
+#else
+#define DLLEXPORT
+#endif
+
+
 // Julia headers (for initialization and gc commands)
 #include "julia.h"
 #include "julia_init.h"
 #include "uv.h"
 
-void setup_args(int argc, char **argv) {
+static void setup_args(int argc, char **argv) {
     uv_setup_args(argc, argv);
     jl_parse_opts(&argc, &argv);
 }
 
-const char *get_sysimage_path(const char *libname) {
+static const char *get_sysimage_path(const char *libname) {
     if (libname == NULL) {
         jl_error("julia: Specify `libname` when requesting the sysimage path");
         exit(1);
@@ -40,7 +47,7 @@ const char *get_sysimage_path(const char *libname) {
     return libpath;
 }
 
-void set_depot_load_path(const char *root_dir) {
+static void set_depot_load_path(const char *root_dir) {
 #ifdef _WIN32
     char *path_sep = ";";
     char *julia_share_subdir = "\\share\\julia";
@@ -83,7 +90,7 @@ void set_depot_load_path(const char *root_dir) {
     free(new_depot_path);
 }
 
-void init_julia(int argc, char **argv) {
+DLLEXPORT void init_julia(int argc, char **argv) {
     setup_args(argc, argv);
 
     const char *sysimage_path = get_sysimage_path(JULIAC_PROGRAM_LIBNAME);
@@ -114,4 +121,4 @@ void init_julia(int argc, char **argv) {
     free(_sysimage_path);
 }
 
-void shutdown_julia(int retcode) { jl_atexit_hook(retcode); }
+DLLEXPORT void shutdown_julia(int retcode) { jl_atexit_hook(retcode); }

--- a/src/juliaconfig.jl
+++ b/src/juliaconfig.jl
@@ -40,7 +40,18 @@ function ldflags()
     fl = "-L$(shell_escape(julia_libdir())) -L$(shell_escape(julia_private_libdir()))"
     if Sys.iswindows()
         fl = fl * " -Wl,--stack,8388608"
-        fl = fl * " -Wl,--export-all-symbols"
+        if VERSION >= v"1.11"
+            # This should not really be necessary if users are using DLLEXPORT as required
+            # on Windows in any compiled / init `.c` files, but we keep these flag for
+            # backwards-compatibility on v1.11+ where it is mostly harmless now that LLVM
+            # 16+ emits `-exclude-symbols` directives
+            #
+            # On Julia 1.10, users must annotate their code with DLLEXPORT for it to link,
+            # since this flag would cause Julia to easily hit COFF symbol limits.
+            #
+            # (see https://github.com/JuliaLang/julia/pull/59736)
+            fl = fl * " -Wl,--export-all-symbols"
+        end
     elseif Sys.islinux()
         fl = fl * " -Wl,--export-dynamic"
     end

